src/components/network/network-loader.js

```diff
@@ -19,117 +19,286 @@ genotet.utils.inherit(genotet.NetworkLoader, genotet.ViewLoader);
+      this.data.network.edges.push(edge);
+    }.bind(this));
+  }.bind(this), 'cannot add genes');
 };
 
 /**
- * Updates the regex that selects the current gene set. Because of gene
- * removal, unlike gene addition, it is hard to incrementally modify the regex.
- * This function sets the gene regex to a verbose concatenation of the gene
- * ids. This is only called in removeGenes_().
+ * Delete one gene from graph and related edges.
+ * @param {!Array<string>} genes Names of the genes.
  * @private
  */
-genotet.NetworkLoader.prototype.updateGeneRegex_ = function() {
-  var regex = '';
-  this.data.nodes.forEach(function(node, index) {
-    regex += node.id + (index == this.data.nodes.length - 1 ? '' : '|');
-  }, this);
-  this.data.geneRegex = regex;
+genotet.NetworkLoader.prototype.deleteGene_ = function(genes) {
+  var geneMap = {};
+  genes.forEach(function(gene) {
+    geneMap[gene] = true;
+  });
+  // delete the nodes
+  for (var i = 0; i < this.data.network.nodes.length; i++) {
+    if (this.data.network.nodes[i].id in geneMap) {
+      this.data.network.nodes.splice(i, 1);
+    }
+  }
+  // delete the edges
+  for (var i = this.data.network.edges.length - 1; i >= 0; i--) {
+    if (this.data.network.edges[i].source in geneMap ||
+      this.data.network.edges[i].target in geneMap) {
+      this.data.network.edges.splice(i, 1);
+    }
+  }
 };
 
 /**
- * Fetches the incident edges of a given node.
- * @param {!Object} node Node of which the incident edges are queried.
+ * Adds one edge to the graph.
+ * @param {!Object} data Input data for adding an edge.
  */
-genotet.NetworkLoader.prototype.incidentEdges = function(node) {
-  var params = {
-    type: 'incident-edges',
-    fileName: this.data.fileName,
-    gene: node.id
-  };
-  this.get(genotet.data.serverURL, params, function(data) {
-    this.data.incidentEdges = data;
-    this.signal('incidentEdges');
-  }.bind(this), 'cannot get incident edges');
+genotet.NetworkLoader.prototype.addOneEdge = function(data) {
+  var source = data.source;
+  var target = data.target;
+  var weight = data.weight;
+  var sourceExists = false, targetExists = false;
+  this.data.network.nodes.forEach(function(node) {
+    if (source == node.id) {
+      sourceExists = true;
+    }
+    if (target == node.id) {
+      targetExists = true;
+    }
+  });
+  if (!sourceExists) {
+    this.data.network.nodes.push({
+      id: source,
+      label: source,
+      isTF: true
+    });
+  }
+  if (!targetExists) {
+    var isTF = false;
+    this.data.networkInfo.nodes.forEach(function(node) {
+      if (target == node.id) {
+        isTF = node.isTF;
+      }
+    });
+    this.data.network.nodes.push({
+      id: target,
+      label: target,
+      isTF: isTF
+    });
+  }
+  this.data.network.edges.push({
+    id: source + ',' + target,
+    source: source,
+    target: target,
+    weight: weight
+  });
+};
+
+/**
+ * Delete one edge from the graph.
+ * @param {!Object} data Input data for delete.
```

Comment:
Be specific on Object. Right now input data is vaguely defined and it's hard to see how to re-use this function.